services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java

```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
     	principalId is self and have MEMBERSHIP permission on Team, or
     	have MEMBERSHIP permission on Team and membership request has been created (but not yet accepted) for principalId
 	 * @param userInfo
 	 * @param teamId the ID of the team
-	 * @param principalId the ID of the one to be added to the team
 	 * @return
 	 */
 	public AuthorizationStatus canAddTeamMember(UserInfo userInfo, String teamId, UserInfo principalUserInfo, boolean alreadyInTeam) throws NotFoundException {
+		// admin can always accept membership
 		if (userInfo.isAdmin()) return AUTHORIZED_ADD_TEAM_MEMBER;
-		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR;
+		if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
+
 		String principalId = principalUserInfo.getId().toString();
-		boolean principalIsSelf = userInfo.getId().toString().equals(principalId);
+		boolean principalIsSelf = userInfo.equals(principalUserInfo);
 		boolean amTeamAdmin = authorizationManager.canAccess(userInfo, teamId, ObjectType.TEAM, ACCESS_TYPE.TEAM_MEMBERSHIP_UPDATE).isAuthorized();
 		long now = System.currentTimeMillis();
+		AuthorizationStatus hasInviteOrRequestToJoin = null;
 		if (principalIsSelf) {
-			// trying to add myself to Team.  
-			if (amTeamAdmin) return AUTHORIZED_ADD_TEAM_MEMBER;
-			// if the team is open, I can join
-			Team team = teamDAO.get(teamId);
-			if (team.getCanPublicJoin()!=null && team.getCanPublicJoin()==true) return AUTHORIZED_ADD_TEAM_MEMBER;
-			// if I'm not a team admin and the team is not open, then I need to have an open invitation
-			if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
-			long openInvitationCount = membershipInvitationDAO.getOpenByTeamAndUserCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openInvitationCount>0L ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
+			// check that principalUserInfo has an invitation to this team or that this team is public
+			hasInviteOrRequestToJoin = this.checkCanAddSelf(teamId, amTeamAdmin, principalId, now);
 		} else {
-			// the member to be added is someone other than me
-			if (!amTeamAdmin) return UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC; // can't add someone unless I'm a Team administrator
-			// can't add someone unless they are asking to be added
-			if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
-			long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
+			// check that principalUserInfo has requested to join this team
+			hasInviteOrRequestToJoin = this.checkCanAddOther(teamId, amTeamAdmin, principalId, now);
+		}
+		if (!hasInviteOrRequestToJoin.isAuthorized()) {
+			return hasInviteOrRequestToJoin;
 		}
+		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) {
```

Comment:
// either (1) principalIsSelf and I have an invitation or (2) principal is other, I am a team admin and there is a membership request